Add extensions-sdk watch build option

moeriki opened the pull request:
I added this to my fork because it's very convenient.

![CleanShot 2021-09-06 at 15 23 27](https://user-images.githubusercontent.com/38758/132224055-f5c74549-69b6-4293-8fee-f0f33621a30a.gif)

Moeriki commented:
Small note. The package seems to have been renamed from `extension-sdk` to `extensions-sdk` in rc92 but I couldn't find this in the release notes.

Moeriki commented:
Fixed an issue with swallowed build errors.

<img width="1136" alt="CleanShot 2021-09-06 at 16 44 40@2x" src="https://user-images.githubusercontent.com/38758/132234114-13657c0c-ca88-4ed5-81c1-59cffd518e78.png">

I ended up lifting quite a bit of code from the Rollup CLI error handler.

https://github.com/rollup/rollup/blob/181f379903988819c79ed6aac7fb5af2fa35877f/cli/logging.ts#L14

Please let me know what you think.

nickrum commented:
Thanks for working on this. Seems like a very usefull addition 👍

We should simplify the error message and re-use the logic for the regular build mode.

There is currently a weird issue where rollup won't rebuild the extensions if there was a syntax error in a `.vue` file, that has been fixed without doing any other changes. Not sure if this is something on our end or an issue with `rollup-plugin-vue`.

Moeriki commented:
I reused the error handler for the build and simplified the error handling just a bit.

Besides the `error.url` which I don't know what it is, I find all the information logged very useful. I just deconstructed and simplified the code.

nickrum commented on packages/extensions-sdk/src/cli/commands/build.ts:
```diff
@@ -75,13 +82,47 @@ export default async function build(options: BuildOptions): Promise<void> {
+					break;
+				}
+				case 'BUNDLE_END':
+					await event.result.write(rollupOutputOptions);
```
This line can be removed as `event.result.write()` is only meant for generating additional bundles.

nickrum commented on packages/extensions-sdk/src/cli/commands/build.ts:
```diff
@@ -75,13 +82,47 @@ export default async function build(options: BuildOptions): Promise<void> {
+				case 'START':
+					spinner.text = 'Building Directus extension...';
+					break;
+				case 'END':
```
This stray switch case can be removed.

nickrum commented on packages/extensions-sdk/src/cli/commands/build.ts:
```diff
@@ -75,13 +82,47 @@ export default async function build(options: BuildOptions): Promise<void> {
+					spinner.succeed(chalk.bold('Done'));
+					spinner.start(chalk.bold('Watching files for changes...'));
+					break;
+				case 'START':
```
Let's use `BUNDLE_START` here for consistency.

nickrum commented on packages/extensions-sdk/src/cli/commands/build.ts:
```diff
@@ -100,7 +141,7 @@ function getRollupOptions(
 				...plugins,
 				nodeResolve({ browser: true }),
-				commonjs({ esmExternals: true, sourceMap: false }),
+				commonjs({ esmExternals: true, sourceMap: true }),
```
Curious to hear what advantages there are of enabling source map generation by the commonjs plugin.

Moeriki replied:
Oops that's committed by mistake.

I was looking to get sourcemaps of my modules to get a stacktrace when running a module within Directus. Without success.

Commit:
Add extension-sdk watch build option
